src/pip/_vendor/tenacity/__init__.py

```diff
@@ -0,0 +1,486 @@
+# -*- coding: utf-8 -*-
+# Copyright 2016-2018 Julien Danjou
+# Copyright 2017 Elisey Zanko
+# Copyright 2016 Étienne Bersac
+# Copyright 2016 Joshua Harlow
+# Copyright 2013-2014 Ray Holder
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+try:
+    import asyncio
+except ImportError:
+    asyncio = None
+
+try:
+    import tornado
+except ImportError:
+    tornado = None
+
+import sys
+import threading
+from concurrent import futures
```

Comment:
This import is causing Python 2.7 builds to break. Seems `tenacity` [requires `futures` and also `monotonic`](https://github.com/jd/tenacity/blob/master/setup.cfg#L26-L27).

Is this a deal-breaker?

Reply:
> Is this a deal-breaker?

That comes down to:

- How big are they?
- How pervasive is the use of these in tenacity?

If it's too big, that's another point in favor of the discussion happening at #7413 w.r.t. Python 2.

Reply:
[`monotonic`](https://github.com/atdt/monotonic) is a single file and [`futures`](https://github.com/agronholm/pythonfutures) is pretty small. Seems `tenacity` does rely on  `futures` quite a bit. I'll add them both.

Reply:
`futures` can be problematic (based on experience using it in Pipenv) because it *only* works on Python 2. You’ll need to take extra care making the code compatible.

Reply:
Yep, I was planning on writing a conditional import on `tenacity` but [I'm not really sure if I can/should change the code](https://github.com/pypa/pip/pull/7592#pullrequestreview-342381722).

Reply:
pip uses `vendoring` to manage patches to the vendor library. The procedure to produce a patch is under-documented, but broadly speaking you should produce a patch in `tools/automation/vendoring/patches` with `git diff [path-to-patched-lib] >`.

Reply:
Yea, to vendor libraries, you'd want to update `vendor.txt` in pip's `_vendor` directory. After that, if there's additional changes needed, we'd want to patch using a `.patch` file as described by @uranusjr above.